src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp

```diff
@@ -0,0 +1,207 @@
+#include <Storages/MergeTree/MergeTreeBackgroundExecutor.h>
+
+#include <algorithm>
+
+#include <Common/setThreadName.h>
+#include <Storages/MergeTree/BackgroundJobsAssignee.h>
+
+
+namespace DB
+{
+
+
+String MergeTreeBackgroundExecutor::toString(Type type)
+{
+    switch (type)
+    {
+        case Type::MERGE_MUTATE:
+            return "MergeMutate";
+        case Type::FETCH:
+            return "Fetch";
+        case Type::MOVE:
+            return "Move";
+    }
+    __builtin_unreachable();
+}
+
+
+void MergeTreeBackgroundExecutor::updateConfiguration()
+{
+    auto new_threads_count = std::max<size_t>(1u, threads_count_getter());
+    auto new_max_tasks_count = std::max<size_t>(1, max_task_count_getter());
+
+    try
+    {
+        pending.set_capacity(new_max_tasks_count);
+        active.set_capacity(new_max_tasks_count);
```

Comment:
Form boost comments:

> If the current number of elements stored in the `circular_buffer` is greater than the desired
new capacity then number of `[size() - new_capacity]` **last** elements will be removed and
the new size will be equal to `new_capacity`.

let's maybe never decrease capacity? (or make it less then size())

Reply:
Deleted dynamic reloading.